src/main/java/com/epam/controller/TraineeController.java

```diff
@@ -0,0 +1,141 @@
+        Trainee updatedTrainee = updatedTraineeOptional.get();
+        mapper.updateEntity(updatedTrainee, request);
+
+        Trainee trainee = traineeService.updateTrainee(updatedTrainee).get();
+        TraineeUpdateResponseDto updateResponseDto = mapper.traineeToTraineeUpdateResponseDto(trainee);
+        return ResponseEntity.ok(updateResponseDto);
+    }
+
+    @DeleteMapping
+    public ResponseEntity<String> deleteTraineeProfile(@RequestParam String username) {
+        boolean deleted = traineeService.deleteTraineeByUsername(username);
+
+        if (deleted) {
+            return ResponseEntity.ok("Trainee profile deleted successfully");
+        } else {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Trainee not found");
+        }
+    }
+
+    @GetMapping("get-not-assigned-trainers")
+    ResponseEntity<List<TrainerProfile>> getNotAssignedTrainers(@RequestParam(value = "username") String username) {
+        List<Trainer> notAssignedTrainers = traineeService.getNotAssignedTrainers(username);
+        List<TrainerProfile> trainerProfiles = mapper.TrainersToTrainerProfiles(notAssignedTrainers);
+        return ResponseEntity.ok(trainerProfiles);
+    }
+
+    @PutMapping("update-by-trainers")
+    public ResponseEntity<List<TrainerProfile>> updateTraineesTrainerList(@RequestBody @Valid TraineeWithTrainersList traineeWithTrainers) {
+        Optional<Trainee> traineeOptional = traineeService.getTraineeByUsername(traineeWithTrainers.getTraineeUsername());
+        if (traineeOptional.isEmpty()) {
+            return ResponseEntity.notFound().build();
+        }
+        Trainee trainee = traineeOptional.get();
+        List<Trainer> trainers = traineeWithTrainers.getTrainersList()
+                .stream()
+                .map(trainerUsername -> getTrainer(trainerUsername, trainee))
+                .collect(Collectors.toList());
+        trainee.getTrainers().addAll(trainers);
+
+        traineeService.updateTrainee(trainee);
+        return ResponseEntity.ok(mapper.TrainersToTrainerProfiles(trainers));
+    }
+
+    private Trainer getTrainer(String trainerUsername, Trainee trainee) {
+        Optional<Trainer> trainerOptional = trainerService.getTrainerByUsername(trainerUsername);
+        if (trainerOptional.isEmpty()) {
+            throw new NoSuchElementException("there is no trainer with username :" + trainerUsername);
+        }
+        Trainer trainer = trainerOptional.get();
+        trainer.getTrainees().add(trainee);
+        return trainer;
+    }
+
+    @GetMapping("/trainings")
+    public ResponseEntity<List<TraineeTrainingResponse>> getTraineeTrainings(@Valid TraineeWithTraining traineeWithTraining) {
+        List<Training> trainings = traineeService.getTraineeTrainingsList(traineeWithTraining);
+
+        return ResponseEntity.ok(mapper.trainingsToTraineeTrainingResponse(trainings));
+    }
+
+    @PatchMapping("/activate")
+    ResponseEntity<String> activateTrainee(String username) {
```

Comment:
Btw, does this method really work? I mean there is no @RequestParam or @PathVariable here.